lib/src/importer/node_package.dart

```diff
@@ -0,0 +1,382 @@
+      return null;
+    }
+
+    var packageRoot = _resolvePackageRoot(packageName, basePath);
+
+    if (packageRoot == null) return null;
+    var jsonPath = p.join(packageRoot, 'package.json');
+
+    var jsonString = readFile(jsonPath);
+    Map<String, dynamic> packageManifest;
+    try {
+      packageManifest = json.decode(jsonString) as Map<String, dynamic>;
+    } catch (e) {
+      throw "Failed to parse $jsonPath for \"pkg:$packageName\": $e";
+    }
+
+    if (_resolvePackageExports(
+            packageRoot, subpath, packageManifest, packageName)
+        case var resolved?) {
+      if (validExtensions.contains(p.extension(resolved))) {
+        return p.toUri(p.canonicalize(resolved));
+      } else {
+        throw "The export for '${subpath ?? "root"}' in "
+            "'$packageName' resolved to '${resolved.toString()}', "
+            "which is not a '.scss', '.sass', or '.css' file.";
+      }
+    }
+    // If no subpath, attempt to resolve `sass` or `style` key in package.json,
+    // then `index` file at package root, resolved for file extensions and
+    // partials.
+    if (subpath == null) {
+      var rootPath = _resolvePackageRootValues(packageRoot, packageManifest);
+      return rootPath != null ? p.toUri(p.canonicalize(rootPath)) : null;
+    }
+
+    // If there is a subpath, attempt to resolve the path relative to the
+    // package root, and resolve for file extensions and partials.
+    var subpathInRoot = p.join(packageRoot, subpath);
+    return FilesystemImporter.cwd.canonicalize(p.toUri(subpathInRoot));
+  }
+
+  @override
+  ImporterResult? load(Uri url) => FilesystemImporter.cwd.load(url);
+
+  /// Splits a [bare import
+  /// specifier](https://nodejs.org/api/esm.html#import-specifiers) `specifier`
+  /// into its package name and subpath, if one exists.
+  ///
+  /// Because this is a bare import specifier and not a path, we always use `/`
+  /// to avoid invalid values on non-Posix machines.
+  (String, String?) _packageNameAndSubpath(String specifier) {
+    var parts = p.url.split(specifier);
+    var name = p.fromUri(parts.removeAt(0));
+
+    if (name.contains('%')) {
+      throw "pkg: name $name must not contain a '%'.";
+    }
+
+    if (name.startsWith('@')) {
+      if (parts.isEmpty) {
+        throw "pkg: name $name is an invalid package name."
+            "Scoped packages, which start with '@', must have a second segment.";
+      }
+      name = p.url.join(name, parts.removeAt(0));
+    }
+    var subpath = parts.isNotEmpty ? p.fromUri(p.url.joinAll(parts)) : null;
+    return (name, subpath);
+  }
+
+  /// Returns an absolute path to the root directory for the most proximate
+  /// installed `packageName`.
+  ///
+  /// Implementation of `PACKAGE_RESOLVE` from the [Resolution Algorithm
+  /// Specification](https://nodejs.org/api/esm.html#resolution-algorithm-specification).
+  String? _resolvePackageRoot(String packageName, String basePath) {
+    String? baseDirectory;
```

Comment:
Why not declare this as `var baseDirectory = basePath`, since it gets `??`ed with `basePath` anyway immediately afterwards?